Experimental/src/Ranges/ImmutableRangeList.cs

```diff
@@ -0,0 +1,273 @@
+            if (ReferenceEquals(this, obj)) return true;
+            if (obj.GetType() != GetType()) return false;
+            return Equals((ImmutableRangeList<TValue>) obj);
+        }
+
+        public override int GetHashCode()
+        {
+            // ReSharper disable NonReadonlyMemberInGetHashCode
+            if (!_hashCode.HasValue)
+                _hashCode = CalculateHashCode();
+
+            return _hashCode.Value;
+            // ReSharper restore NonReadonlyMemberInGetHashCode
+        }
+
+        private int CalculateHashCode()
+        {
+            if (Values == null)
+                return 0;
+            int res = 0x2D2816FE;
+            for (int i = 0; i < Values.Count; i++)
+            {
+                res = res * 31 + Values[i].GetHashCode();
+            }
+            return res;
+        }
+
+        public string DisplayValue()
+        {
+            string result;
+
+            if (IsEmpty)
+            {
+                result = string.Empty;
+            }
+            else if (IsFull)
+            {
+                result = "...";
+            }
+            else
+            {
+                result = string.Join(", ", Values.Select(v => v.DisplayValue()));
+            }
+
+            return "[" + result + "]";
+        }
+
+        public ImmutableRangeList<TValue> Add(Range<TValue> range)
+        {
+            if (range.IsEmpty)
+                return this;
+
+            if (Values == null)
+                return new ImmutableRangeList<TValue>(range);
+
+            return this.Concat(new[] {range}).ToRangeList();
+        }
+
+        public ImmutableRangeList<TValue> AddRange(IEnumerable<Range<TValue>> ranges)
+        {
+            if (ranges == null)
+                return this;
+
+            return Values == null ? ranges.ToRangeList() : Values.Concat(ranges).ToRangeList();
+        }
+
+        public ImmutableRangeList<TValue> RemoveRange(IEnumerable<Range<TValue>> ranges)
+        {
+            if (Values == null)
+                return this;
+
+            if (ranges == null)
+                return this;
+
+            var rangesList = ranges.ToList();
+
+            var sourceList = Values;
+            List<Range<TValue>> copyValues = null;
+
+            foreach (var toRemove in rangesList)
+            {
+                int i = 0;
+                do
+                {
+                    var current = sourceList[i];
+                    if (current.Intersects(toRemove))
+                    {
+                        var removed = current.Exclude(toRemove);
+                        if (copyValues == null)
+                        {
+                            copyValues = Values.ToList();
+                            sourceList = copyValues;
+                        }
+                        sourceList.RemoveAt(i);
+                        var prevCount = sourceList.Count;
+                        sourceList.InsertRange(i, removed);
```

Comment:
Is a range order is really important here? If not, it would be better to postpone all insertions until the end of the current iteration (using a temp list)  and then insert everything at the end of the sourceList